### PR TITLE
[HNSW][PQ/BQ] Prefill Compressed vector cache in parallel

### PR DESCRIPTION
**Pull request by etiennedi**
### What's being changed:

In #5468, we introduced a parallel iterator that can prefill the vector index cache much faster using SSDs by parallelizing the loading. However, this iterator was only used for the flat index when prefilling BQ vectors. This PR makes the iterator generic, so it can be used with PQ and BQ. In addition, it is now used for HNSW+BQ/PQ.

The test was extended to also verify the correctness of the extracted vectors.

In addition, a progress logger was added, so we have more observability into how fast things are loading on large setups:

image

## Recall Sanity checks

The idea is that if the recall is the same as before, we know that everything was loaded correctly.

#### Control (`stable/v1.25`)
image

#### This branch
image

#### Control (`stable/v1.25`)

image

#### This branch

image

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

**Review by aliszka — DISMISSED** [review]

File: adapters/repos/db/vector/hnsw/startup.go [hunk 1]
```diff
@@ -231,7 +231,7 @@ func (h *hnsw) prefillCache() {
 
 		h.logger.WithFields(logrus.Fields{
 			"action":   "prefill_cache",
-			"duration": 60 * time.Minute,
+			"duration": 240 * time.Minute,
```
[the remark this answers is not in the document]
Comment (etiennedi) [reply adapters/repos/db/vector/hnsw/startup.go:234]: Nope, thanks, I will revert.

**Review by abdelr — DISMISSED** [review]

File: adapters/repos/db/vector/compressionhelpers/compression.go [hunk 2]
```diff
@@ -216,61 +217,50 @@ func (compressor *quantizedVectorsCompressor[T]) initCompressedStore() error {
...
-				"len":    len(v),
-				"lenk":   len(k),
-			}).Warn("skipping compressed vector with unexpected length")
-			continue
-		}
+	for v := range channel {
+		vecs = append(vecs, v...)
+	}
 
+	count := 0
```
[the remark this answers is not in the document]
Comment (etiennedi) [reply adapters/repos/db/vector/compressionhelpers/compression.go:245]: yup

File: adapters/repos/db/vector/compressionhelpers/parallel_iterator.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+func (cpi *parallelIterator[T]) IterateAll() chan []VecAndID[T] {
+	if cpi.parallel <= 1 {
+		// caller explicitly wants no parallelism, fallback to regular cursor
+		return cpi.iterateAllNoConcurrency()
+	}
+
+	stopTracking := cpi.startTracking()
+	defer func() {
+	}()
```
[the remark this answers is not in the document]
Comment (etiennedi) [reply adapters/repos/db/vector/compressionhelpers/parallel_iterator.go:66]: The best function ever, but it was written in invisible ink 🤣 

I'll remove it...

File: adapters/repos/db/vector/compressionhelpers/quantizer.go [hunk 4]
```diff
@@ -72,6 +78,29 @@ func (bq *BinaryQuantizer) FromCompressedBytes(compressed []byte) []uint64 {
...
 
+// FromCompressedBytesWithSubsliceBuffer is like FromCompressedBytes, but
+// instead of allocating a new slice you can pass in a buffer to use. It will
+// slice something off of that buffer. If the buffer is too small, it will
+// allocate a new buffer.
+func (bq *BinaryQuantizer) FromCompressedBytesWithSubsliceBuffer(compressed []byte, buffer *[]uint64) []uint64 {
+	l := len(compressed) / 8
+	if len(compressed)%8 != 0 {
+		l++
+	}
```
[the remark this answers is not in the document]
Comment (etiennedi) [reply adapters/repos/db/vector/compressionhelpers/quantizer.go:89]: @abdelr can say more, I copy/pasted this from the existing logic. It think what this doing is essentially creating a new uint64 container if the original len is not divisible by 8.

E.g.
* `len(in)==8` -> `l==1`
* `len(in)==15` -> line 86 would now return `1`, but we can't fit 15 bytes of information into just one `uint64`, so we need another one.

File: adapters/repos/db/vector/compressionhelpers/quantizer.go [hunk 5]
```diff
@@ -72,6 +78,29 @@ func (bq *BinaryQuantizer) FromCompressedBytes(compressed []byte) []uint64 {
...
+
+	if len(*buffer) < l {
+		*buffer = make([]uint64, 1000*l)
```
[the remark this answers is not in the document]
Comment (etiennedi) [reply adapters/repos/db/vector/compressionhelpers/quantizer.go:92]: I assume this resolved based on the slack conversation? 

For those reading here:

* This is an optimization to allocate one big buffer which is then sliced into many small buffers. This is about 3x faster in a micro-benchmark and still ~10% faster in some e2e cases
* We identified a small flaw in that logic that could waste a lot of memory that would never have been freed
* @dirkkul pushed some more commits that now make sure we have the best of both worlds: A single big allocation (faster), but without the waste.

File: adapters/repos/db/vector/compressionhelpers/quantizer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply adapters/repos/db/vector/compressionhelpers/quantizer.go:89]: oki, makes sense!

File: adapters/repos/db/vector/compressionhelpers/parallel_iterator_test.go [hunk 6]
```diff
@@ -69,7 +74,7 @@ func TestCompressedParallelIterator(t *testing.T) {
 		},
 	}
 
-	quantization := []string{"pq"}
+	quantization := []string{"pq", "bq"}
```
Comment (etiennedi) [inline adapters/repos/db/vector/compressionhelpers/parallel_iterator_test.go:77]: Wow! Good catch. Running both was the whole reason I added this. Never realized we were only running one. Great to see that tests still pass!

**Comment by etiennedi** [comment]
We made quite a few changes since the original PR, so I ran another recall sanity check. The following shows two runs. The first was a query run immediately after importing (i.e. the preloading logic never ran). Then I restarted the server, forcing a parallel reload. Recall is identical ✅ !

image

**Review by antas-marcin — APPROVED** [review]

**Pull request merged**